api/api.go

```diff
@@ -211,9 +213,9 @@ func (a *API) Retrieve(ctx context.Context, addr storage.Address) (reader storag
 }
 
 // Store wraps the Store API call of the embedded FileStore
-func (a *API) Store(ctx context.Context, data io.Reader, size int64, toEncrypt bool) (addr storage.Address, wait func(ctx context.Context) error, err error) {
+func (a *API) Store(ctx context.Context, data io.Reader, size int64, toEncrypt bool, pinCounter uint8) (addr storage.Address, wait func(ctx context.Context) error, err error) {
```

Comment:
> @jmozah a few notes about this PR:
> 
> 1. the delta is huge (61 files so far). we'd have to break this down to smaller units

Actually the real changes are very few.. because of the signature changes in upload APIs.. this affected large number of files..

> 2. there's no test coverage whatsoever for the new functionality

This is WIP PR.. Test cases are in the making.. :-)

> 3. the necessity of `pinCounter` (which is always set to `1` in the api) is probably arguable

Pin counter is not always 1. When you upload a file with --pin flag, the pin counter is set to 1 otherwise 0. When i add support for "swarm pin <file hash>" command.. then the pin counter will increment for every pin.

> 4. `pinCounter` definitely should not be `uint8`. Imagine that within a certain pin you limit yourself to 255 instances of the same chunk, which is very low (just imagine all of the `.DS_Store` files and other possible examples of just 32 byte repetitions)

hmmm... I will make it uint64.

> 5. i'm not sure i understand why we would need to store information about whether to traverse a manifest as part of the localstore index. it seems like to much of high-level information to store somewhere

Well. We need to know if a hash is a RAW file hash or a manifest hash. When we unpin a pinned hash, we need to traverse all the chunks underneath it to unpin it. If the hash was a file.. it is one way and if the hash is a manifest, we need to parse the manifest and all the files under the manifest to unpin all the chunks.  

The only other way i see is to read the contents of the hash and try to JSONUnMarshel it... if it fails.. it is a file.. otherwise it is a manifest..  This didn't look so nice... Thats why storing the raw in the index.


>    i think that a good next step would be to discuss the correct test vectors for this functionality and proceed from there

Reply:
Change to uint64